### ISSUE TITLE
Restic Performance limited to 1.7Gbit / 4 Cores on 32 core system

### ISSUE THREAD

**Issue by ravench** [restic/restic#4477]
Output of `restic version`
--------------------------
`restic 0.16.0 compiled with go1.20.6 on linux/amd64`

How did you run restic exactly?
-------------------------------

Comands:
```shell

restic -r dest_restic backup --pack-size 128 source_restic/
restic -r dest_restic backup --pack-size 128 --no-scan -o local.connections=8 --read-concurrency 10 source_restic/
```

Example Output: 
```plain
enter password for repository:
repository 42ff0d97 opened (version 2, compression level auto)
using parent snapshot 7f089c79

Files:           1 new,     0 changed,     3 unmodified
Dirs:            0 new,     3 changed,     0 unmodified
Added to the repository: 101.236 GiB (101.003 GiB stored)

processed 4 files, 296.661 GiB in 10:45
snapshot ae4cfc77 saved
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Local NVMe

Expected behavior
-----------------

Use all available resources on the system / produce a load similar to the number of physical cores (32 in this case) or a throughput at the limit of the network or SSD speed. (between 5 and 8 Gigabits.)

Actual behavior
---------------

Restic will not use more than around 400% of CPU and do a max throughput of around 1.7Gbit.

Steps to reproduce the behavior
-------------------------------

- Run Restic on a decent server (32C / 600G of RAM) with local NVMe Storage.
- Create a local source folder with some Files between 10GB and 160GB and a local repo.
- Run restic with no options and the above options.
- You should see that there is some bottleneck which is neither the CPU nor the storage.

Do you have any idea what may have caused this?
-----------------------------------------------

Not really. 
Restic seems to use all 64 threads, but in total they only generate around 350-400% of CPU load. Most threads seem fairly idle, at around 5% CPU usage.

It's not the SSD, (Samsung 980 Pro) which can write out at least 7.5Gbit/s: 
(Note: In this test the limiting factor was the one CPU Core dd runs on, the SSD datasheet claims speeds up to 5GB/s (40Gbit/s))

```shell
dd if=/dev/zero of=testfile1.img bs=512M count=200
200+0 records in
200+0 records out
107374182400 bytes (107 GB, 100 GiB) copied, 105.957 s, 1.0 GB/s

sudo sysctl vm.drop_caches=3

dd if=testfile1.img of=testfile2.img bs=512M count=200
200+0 records in
200+0 records out
107374182400 bytes (107 GB, 100 GiB) copied, 139.81 s, 768 MB/s


# reading from urandom is slow
dd if=/dev/urandom of=testfile3.img bs=512M count=200
200+0 records in
200+0 records out
107374182400 bytes (107 GB, 100 GiB) copied, 377.704 s, 284 MB/s

sudo sysctl vm.drop_caches=3

 dd if=testfile3.img of=testfile4.img bs=512M count=200
200+0 records in
200+0 records out
107374182400 bytes (107 GB, 100 GiB) copied, 143.141 s, 750 MB/s

```

In the end I want to back up from one remote system (NetApp) to another (Synology), so I also tested with various network setups.
I tried backups from NFS to local, local to NFS and NFS to NFS directly, and got almost exactly the same results as local to local. (SFTP wasn't faster, I tried that before). 

When using `cp` to copy around 300GBytes from a NFS mount to local disk I got around 4.5Gbits of throughput. (On a 10G network)

Our experts tell me that the 1.7 Gbits I'm seeing from restic when using network backends (same speed / execution time as local to local) is well below what AES offloading in a modern CPU is able to handle. 

So basically this really seems to be a limitation in restic and not in the hardware used.

Do you have an idea how to solve the issue?
-------------------------------------------

Not really, I'm afraid.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

We found this issue while testing against similar tools and restic clearly outperformed them.

**Comment by ravench** [comment]
This might relate to <https://forum.restic.net/t/support-parallelism-when-backing-up-very-large-files/4141>

**Comment by ravench** [comment]
Some more tests with ramdisks:

```shell


root@server:~ # mkdir /tmp/ramdisk1
mkdir: created directory '/tmp/ramdisk1'

root@server:~ # mkdir /tmp/ramdisk2
mkdir: created directory '/tmp/ramdisk2'

root@server:~ # mount -t tmpfs -o size=150G ramdisk1 /tmp/ramdisk1

root@server:~ # mount -t tmpfs -o size=150G ramdisk2 /tmp/ramdisk2

root@server:~ # dd if=/home/testuser/100G_urandom.img of=/tmp/ramdisk1/100G_urandom.img bs=512M
200+0 records in
200+0 records out
107374182400 bytes (107 GB, 100 GiB) copied, 47.3863 s, 2.3 GB/s

root@server:~ # cat /proc/loadavg
0.19 0.11 0.04 1/779 7321

root@server:~ # restic -r /tmp/ramdisk2/restic_repo init --pack-size 128
enter password for new repository:
enter password again:
created restic repository f0a955022e at /tmp/ramdisk2/restic_repo

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.

root@server:~ # restic -r /tmp/ramdisk2/restic_repo/ backup --pack-size 128 /tmp/ramdisk1/ ; cat /proc/loadavg
enter password for repository:
repository f0a95502 opened (version 2, compression level auto)
created new cache in /root/.cache/restic
no parent snapshot found, will read all files

Files:           1 new,     0 changed,     0 unmodified
Dirs:            2 new,     0 changed,     0 unmodified
Added to the repository: 100.004 GiB (100.010 GiB stored)

processed 1 files, 100.000 GiB in 5:46
snapshot 61ed2151 saved
4.49 3.98 2.41 1/776 7703

root@server:~ # rm -r /tmp/ramdisk2/restic_repo
rm: remove 1 argument recursively? y

root@server:~ # restic -r /tmp/ramdisk2/restic_repo init --pack-size 128
enter password for new repository:
enter password again:
created restic repository 4c60eff88b at /tmp/ramdisk2/restic_repo

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.

root@server:~ # restic -r /tmp/ramdisk2/restic_repo/ backup --pack-size 128 --no-scan -o local.connections=8 --read-concurrency 10 /tmp/ramdisk1/ ; cat /proc/loadavg
enter password for repository:
repository 4c60eff8 opened (version 2, compression level auto)
created new cache in /root/.cache/restic
no parent snapshot found, will read all files

Files:           1 new,     0 changed,     0 unmodified
Dirs:            2 new,     0 changed,     0 unmodified
Added to the repository: 100.004 GiB (100.010 GiB stored)

processed 1 files, 100.000 GiB in 5:47
snapshot a39da69f saved
3.99 2.87 1.32 1/775 7419

root@server:~ # rm -r /tmp/ramdisk2/restic_repo
rm: remove 1 argument recursively? y

root@server:~ # restic -r /tmp/ramdisk2/restic_repo init --pack-size 128
enter password for new repository:
enter password again:
created restic repository 5bb25dd8c5 at /tmp/ramdisk2/restic_repo

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.

root@server:~ # restic -r /tmp/ramdisk2/restic_repo/ backup --pack-size 128 --no-scan -o local.connections=16 --read-concurre
ncy 20 /tmp/ramdisk1/ ; cat /proc/loadavg
enter password for repository:
repository 5bb25dd8 opened (version 2, compression level auto)
created new cache in /root/.cache/restic
no parent snapshot found, will read all files

Files:           1 new,     0 changed,     0 unmodified
Dirs:            2 new,     0 changed,     0 unmodified
Added to the repository: 100.004 GiB (100.010 GiB stored)

processed 1 files, 100.000 GiB in 5:45
snapshot 87702100 saved
4.48 3.28 1.52 1/777 7603
 
root@server:~ #
```

`100*8/(5*60+47)` puts us around 2.3Gbit/s which is better, but still not seeing more than ` 4` load, so it seems like a limitation of the amount of CPU power that can be thrown at a problem?

**Comment by ravench** [comment]
Using 100 files of 1GB instead 1 file with 100GB gives significantly better performance, when also increasing  read concurrency.

It seems, that a good default for read concurrency would be `cores/4`.

`100*8/(108)` puts us at around 7.4Gbit/s, which is fine for my purposes. However, still not managing to get 32 CPU cores fully loaded, also when reading 16 files in parallel.

``` shell
root@server:~ # rm -r /tmp/ramdisk*/*
rm: remove 2 arguments recursively? y

root@server:~ # for i in $(seq 1 100);
do
     dd if=/dev/urandom of=/tmp/ramdisk1/1G_urandom_$i bs=512M count=2
done
2+0 records in
2+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 3.45135 s, 311 MB/s
2+0 records in
2+0 records out
(...)


root@server:~ # restic -r /tmp/ramdisk2/restic_repo init --pack-size 128
enter password for new repository:
enter password again:
created restic repository 5a27b7c165 at /tmp/ramdisk2/restic_repo

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.

root@server:~ # du -hs /tmp/ramdisk*
100G    /tmp/ramdisk1
8.0K    /tmp/ramdisk2

root@server:~ # restic -r /tmp/ramdisk2/restic_repo/ backup --pack-size 128 /tmp/ramdisk1/ ; cat /proc/loadavg
enter password for repository:
repository 5a27b7c1 opened (version 2, compression level auto)
created new cache in /root/.cache/restic
no parent snapshot found, will read all files

Files:         100 new,     0 changed,     0 unmodified
Dirs:            2 new,     0 changed,     0 unmodified
Added to the repository: 100.004 GiB (100.010 GiB stored)

processed 100 files, 100.000 GiB in 3:09
snapshot c5086f6a saved
9.13 4.83 2.54 1/776 7915

root@server:~ # rm -r /tmp/ramdisk2/restic_repo/
rm: remove 1 argument recursively? y

root@server:~ # restic -r /tmp/ramdisk2/restic_repo init --pack-size 128
enter password for new repository:
enter password again:
created restic repository 89866798d3 at /tmp/ramdisk2/restic_repo

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.

root@server:~ # restic -r /tmp/ramdisk2/restic_repo/ backup --pack-size 128 --no-scan -o local.connections=8 --read-concurren
cy 8 /tmp/ramdisk1/ ; cat /proc/loadavg
enter password for repository:
repository 89866798 opened (version 2, compression level auto)
created new cache in /root/.cache/restic
no parent snapshot found, will read all files

Files:         100 new,     0 changed,     0 unmodified
Dirs:            2 new,     0 changed,     0 unmodified
Added to the repository: 100.004 GiB (100.010 GiB stored)

processed 100 files, 100.000 GiB in 1:53
snapshot 4f830b87 saved
13.50 7.62 3.93 1/778 8004

root@server:~ # rm -r /tmp/ramdisk2/restic_repo/
rm: remove 1 argument recursively? y

root@server:~ # restic -r /tmp/ramdisk2/restic_repo init --pack-size 128
enter password for new repository:
enter password again:
created restic repository 7ffc7e7c5c at /tmp/ramdisk2/restic_repo

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.

root@server:~ # restic -r /tmp/ramdisk2/restic_repo/ backup --pack-size 128 --no-scan -o local.connections=16 --read-concurrency 16 /tmp/ramdisk1/ ; cat /proc/loadavg
enter password for repository:
repository 7ffc7e7c opened (version 2, compression level auto)
created new cache in /root/.cache/restic
no parent snapshot found, will read all files

Files:         100 new,     0 changed,     0 unmodified
Dirs:            2 new,     0 changed,     0 unmodified
Added to the repository: 100.004 GiB (100.010 GiB stored)

processed 100 files, 100.000 GiB in 1:51
snapshot 8a87c674 saved
11.29 8.40 4.75 1/777 8114

root@server:~ # rm -r /tmp/ramdisk2/restic_repo/
rm: remove 1 argument recursively? y

root@server:~ # restic -r /tmp/ramdisk2/restic_repo init --pack-size 128
enter password for new repository:
enter password again:
created restic repository 82d2bf7fe7 at /tmp/ramdisk2/restic_repo

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.

root@server:~ # restic -r /tmp/ramdisk2/restic_repo/ backup --pack-size 128 --no-scan -o local.connections=8 --read-concurren
cy 32 /tmp/ramdisk1/ ; cat /proc/loadavg
enter password for repository:
repository 82d2bf7f opened (version 2, compression level auto)
created new cache in /root/.cache/restic
no parent snapshot found, will read all files

Files:         100 new,     0 changed,     0 unmodified
Dirs:            2 new,     0 changed,     0 unmodified
Added to the repository: 100.004 GiB (100.010 GiB stored)

processed 100 files, 100.000 GiB in 1:48
snapshot ef6236c7 saved
9.93 8.39 5.26 1/778 8211

root@server:~ #
```

**Comment by solracsf** [comment]
Can `GOGC` and `GOMAXPROCS` help here?
https://concertio.com/blog/optimizing-the-go-garbage-collector-and-concurrency/

**Comment by ravench** [comment]
Setting `GOGC=25600` and `GOMAXPROCS=128` improves transfer speeds, while keeping load roughly constant and using signifficantly more RAM.

Repeating the ramdisk tests from above:

| Options | Files | Execution Time 1 | Load 1 | Execution time 2 | Load 2 | Comment |
|---------|-------|------------------|--------|------------------|--------|---------|
| `--pack-size 128` | 1x100G | 5:46 | 4.49 | 3:48 | 5.57 |  |
| `--pack-size 128 --no-scan -o local.connections=8 --read-concurrency 10` | 1x100G | 5:47 | 3.99 | 3:43 | 4.54 |  |
| `--pack-size 128 --no-scan -o local.connections=16 --read-concurrency 20` | 1x100G | 5:45 | 4.48 | 3:44 | 5.07 |  |
| `--pack-size 128` | 100x1G | 3:09 | 9.13 | 1:53 | 8.43 |  |  
| `--pack-size 128 --no-scan -o local.connections=8 --read-concurrency 8` | 100x1G | 1:53 | 13.50 | 1:26 | 13.20 |  |
| `--pack-size 128 --no-scan -o local.connections=16 --read-concurrency 16` | 100x1G | 1:51 | 11.29 | - | - | failed (out of memory)  |

However, Restic ends up using 300GB of RAM in the last test. Maybe I overdid it a bit..

**Comment by MichaelEischer** [comment]
The data processing pipeline in restic works as follows (numbers based on AMD Ryzen 7 PRO 5850U):

1. the file content is read and cut into smaller chunks (limited at about 450 MB/s per file). Multiple files can be read in parallel, but for each file the chunking happens sequentially.
2. calculate the sha256 hash for each chunk. This step and all following can happen in parallel with each CPU processing a different chunk (for large chunks about 550MB/s, or 2GB/s with CPUs supporting SHA-NI)
3. If a chunk is already stored in the repository, then processing ends here.
4. Compress the chunk (zstd, at default compression level, a few hundred MB/s should be possible)
5. Encrypt the chunk using AES (> 1.4 GB/s with AES-NI)
6. Store the encrypted chunk in the pack file and hash it again

That is, when processing a single large file, then reading & chunking are the bottleneck. For many large files in parallel, the bottleneck shifts. Now, writing at most two data files in parallel is the bottleneck, which is why increasing the `local.connections` helps. The last two test runs which take about 1:50, process 0.9GB/s which seems like a pretty decent speed. Beyond that there are probably a few smaller bottlenecks that add up.

It would definitely be possible to process very large files in e.g. 10GB chunks, such that these can be processed in parallel.

> It seems, that a good default for read concurrency would be `cores/4`.

That is only true for SSD storage. HDDs will most likely suffer from increased concurrency.
